File: project/jimmer-spring-boot-starter/src/test/java/org/babyfish/jimmer/spring/java/dal/BookRepository.java
package org.babyfish.jimmer.spring.java.dal;

import org.babyfish.jimmer.spring.repository.JRepository;
import org.babyfish.jimmer.spring.java.model.*;
import org.babyfish.jimmer.sql.ast.Predicate;
import org.babyfish.jimmer.sql.fetcher.Fetcher;
import org.jetbrains.annotations.Nullable;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.math.BigDecimal;
import java.util.List;

public interface BookRepository extends JRepository<Book, Long> {

    BookTable table = BookTable.$;

    List<Book> findByNameOrderByNameAscEditionDesc(
            @Nullable String name,
            @Nullable Fetcher<Book> fetcher
    );

    Page<Book> findByNameLikeIgnoreCaseAndStoreNameOrderByNameAscEditionDesc(
            Pageable pageable,
            @Nullable Fetcher<Book> fetcher,
            @Nullable String name,
            @Nullable String storeName
    );

    List<BigDecimal> findDistinctPriceByPriceBetween(
            @Nullable BigDecimal min,
            @Nullable BigDecimal max
    );

    default Page<Book> findBooks(
            int pageIndex,
            int pageSize,
            @Nullable String name,
            @Nullable String storeName,
            @Nullable String authorName,
            Fetcher<Book> fetcher
    ) {
        AuthorTableEx author = AuthorTableEx.$;
        return pager(
                pageIndex,
                pageSize
        ).execute(
                sql().createQuery(table)
                        .whereIf(name != null, table.name().ilike(name))
                        .whereIf(storeName != null, table.store().name().ilike(name))
                        .whereIf(
                                authorName != null,
                                table.id().in(
                                        sql().createSubQuery(author)
                                                .where(
                                                        Predicate.or(
                                                                author.firstName().ilike(authorName),
                                                                author.lastName().ilike(authorName)
                                                        )
                                                )
                                                .select(author.books().id())
                                )
                        )
                        .select(table.fetch(fetcher))
        );
    }
}

File: example/kotlin/jimmer-sql-kt/src/main/kotlin/org/babyfish/jimmer/example/kt/sql/dal/BookRepository.kt
package org.babyfish.jimmer.example.kt.sql.dal

import org.babyfish.jimmer.example.kt.sql.model.*
import org.babyfish.jimmer.spring.repository.KRepository
import org.babyfish.jimmer.sql.fetcher.Fetcher
import org.babyfish.jimmer.sql.kt.ast.expression.*
import org.springframework.data.domain.Page

interface BookRepository : KRepository<Book, Long> {

    fun findBooks(
        pageIndex: Int,
        pageSize: Int,
        name: String?,
        storeName: String?,
        authorName: String?,
        fetcher: Fetcher<Book>?
    ): Page<Book> =
        pager(pageIndex, pageSize).execute(
            sql.createQuery(Book::class) {
                name?.takeIf { it.isNotEmpty() }?.let {
                    where(table.name ilike it)
                }
                storeName?.takeIf { it.isNotEmpty() }?.let {
                    where(table.store.name ilike it)
                }
                authorName?.takeIf { it.isNotEmpty() }?.let {
                    where(
                        table.id valueIn subQuery(Author::class) {
                            where(
                                or(
                                    table.firstName ilike it,
                                    table.lastName ilike it
                                )
                            )
                            select(table.books.id)
                        }
                    )
                }
                orderBy(table.name)
                orderBy(table.edition.desc())
                select(table.fetch(fetcher))
            }
        )

    fun findNewestIdsByStoreIds(storeIds: Collection<Long>): Map<Long, Long> =
        sql
            .createQuery(Book::class) {
                where(table.store.id valueIn storeIds)
                groupBy(table.store.id)
                select(
                    table.store.id,
                    max(table.id).asNonNull()
                )
            }
            .execute()
            .associateBy({it._1}) {
                it._2
            }
}